harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -240,106 +518,281 @@ def on_shutdown(callback_name: str, on_trial_shutdown: Callable) -> None:
+        # XXX: is this actually necessary? there's no searcher mandate, but still seems useful to
+        # do a last checkpoint before finish
+        if not self._validation_is_current():
+            self._validate()
+
+        if not self._checkpoint_is_current():
+            self._checkpoint(already_exiting=False)
+
+    def _validate_for_op(self, op: core.SearcherOperation, searcher_length: TrainUnit):
+        val_metrics = self._validate()
+
+        # Validation step complete. Chief must:
+        # Report metrics to searcher API
+        # Report metrics to core API
+        # Checkpoint if policy dictates
+        if self._is_chief:
+            searcher_metric = self._validate_searcher_metric(val_metrics)
+            if self._steps_until_complete(searcher_length) < 1:
+                op.report_completed(searcher_metric)
+
+            if self._ckpt_policy == "best" and not self._checkpoint_is_current():
+                best_validation_before = self._core_context.train.get_experiment_best_validation()
+
+            self._core_context.train.report_validation_metrics(
+                self._state.batches_trained, val_metrics
+            )
+
+            if not self._checkpoint_is_current():
+                if self._ckpt_policy == "all" or (
+                    self._ckpt_policy == "best"
+                    and self._is_best_validation(now=searcher_metric, before=best_validation_before)
+                ):
+                    self._checkpoint(already_exiting=False)
+
+    def _train_for_op(
+        self, op: core.SearcherOperation, searcher_unit: core.Unit, train_steps: List[_TrainStep]
+    ):
+        searcher_length = TrainUnit._from_searcher_unit(op.length, searcher_unit)
+
+        while self._steps_until_complete(searcher_length) > 0:
+            train_steps, training_metrics = self._train_with_steps(
+                self.training_enumerator, train_steps
+            )
+
+            self._report_training_metrics(training_metrics)
+
+            for train_step in train_steps:
+                if not train_step.limit_reached:
+                    continue
+
+                # Train step limits reached, proceed accordingly.
+                # Report metrics and searcher progress before validation/checkpoint
+                # Because of this, no extra logic is needed for scheduling_unit step
+                if not op._completed and self._is_chief:
+                    self._report_searcher_progress(op, searcher_unit)
+
+                if train_step.step_type == _TrainStepType.VALIDATE:
+                    if not self._validation_is_current():
+                        self._validate_for_op(op, searcher_length)
+
+                elif train_step.step_type == _TrainStepType.CHECKPOINT:
+                    if not self._checkpoint_is_current():
+                        self._checkpoint(already_exiting=False)
+
+                # Reset train step limit
+                train_step.limit_reached = False
```

Comment:
do you save train step state in checkpoints?  If not, is there a world where you could validate-checkpoint-pause-activate-validate, without any training in between?

Or if not, is it possible to avoid the `limit_reached` state somehow?

Reply:
the train step state is saved as "last_val" and "last_ckpt". they don't really have any meaningful state outside of this.

Reply:
ok

Reply:
It would be a problem if you checkpointed, validated, crashed, woke back up, validated again (a duplicate) and the crashed reporting duplicate validation metrics to the master.

This could happen if a checkpoint_period-caused checkpoint occurs on the same batch that a searcher op is completed.  I think you can just avoid taking that checkpoint within the `for op in searcher.operations()` loop and let the other checkpoint call take care of it.